File: openc3/lib/openc3/logs/packet_log_reader.rb
# encoding: ascii-8bit

# Copyright 2022 Ball Aerospace & Technologies Corp.
# All Rights Reserved.
#
# This program is free software; you can modify and/or redistribute it
# under the terms of the GNU Affero General Public License
# as published by the Free Software Foundation; version 3 with
# attribution addendums as found in the LICENSE.txt
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU Affero General Public License for more details.

# Modified by OpenC3, Inc.
# All changes Copyright 2024, OpenC3, Inc.
# All Rights Reserved
#
# This file may also be used under the terms of a commercial license
# if purchased from OpenC3, Inc.

require 'openc3/core_ext/io'
require 'openc3/packets/packet'
require 'openc3/packets/json_packet'
require 'openc3/io/buffered_file'
require 'openc3/logs/packet_log_constants'
require 'cbor'

module OpenC3
  # Reads a packet log of either commands or telemetry.
  class PacketLogReader
    include PacketLogConstants

    attr_reader :redis_offset
    attr_reader :last_offsets
    attr_reader :filename

    MAX_READ_SIZE = 1000000000

    # Create a new log file reader
    def initialize
      reset()
    end

    # Yields back each packet as it is found in the log file.
    #
    # @param filename [String] The log file to read
    # @param identify_and_define [Boolean] Once the packet has been read from
    #   the log file, whether to both identify the packet by setting the target
    #   and packet name, and define the packet by populating all the items.
    # @param start_time [Time|nil] Time at which to start returning packets.
    #   Packets found with a timestamp before this time are ignored. Pass nil
    #   to return all packets.
    # @param end_time [Time|nil] Time at which to stop returning packets.
    #   Packets found with a timestamp after this time are ignored. Pass nil
    #   to return all packets.
    # @yieldparam packet [Packet]
    # @return [Boolean] Whether we reached the end_time while reading
    def each(filename, identify_and_define = true, start_time = nil, end_time = nil)
      reached_end_time = false
      open(filename)

      # seek_to_time(start_time) if start_time

      while true
        packet = read(identify_and_define)
        break unless packet

        time = packet.packet_time
        if time
          next if start_time and time < start_time
          # If we reach the end_time that means we found all the packets we asked for
          # This can be used by callers to know they are done reading
          if end_time and time > end_time
            reached_end_time = true
            break
          end
        end
        yield packet
      end
      reached_end_time
    ensure # No implicit return value in the ensure block
      close()
    end

    # @param filename [String] The log filename to open
    # @return [Boolean, Exception] Returns true if successfully changed to configuration specified in log,
    #    otherwise returns false and potentially an Exception class if an error occurred.  If no error occurred
    #    false indicates that the requested configuration was simply not found.
    def open(filename, string_io: nil)
      close()
      reset()
      @filename = filename
      if string_io
        @file = string_io
      else
        @file = BufferedFile.open(@filename, 'rb')
      end
      @max_read_size = @file.size
      @max_read_size = MAX_READ_SIZE if @max_read_size > MAX_READ_SIZE
      return read_file_header()
    rescue => e
      close()
      raise e
    end

    # Closes the current log file
    def close
      @file.close if @file and !@file.closed?
      @filename = nil
    end

    # Read a packet from the log file
    #
    # @param identify_and_define (see #each)
    # @return [Packet]
    def read(identify_and_define = true)
      # Read entry length
      length = @file.read(4)
      return nil if !length or length.length <= 0

      length = length.unpack('N')[0]
      entry = @file.read(length)
      flags = entry[0..1].unpack('n')[0]

      cmd_or_tlm = :TLM
      cmd_or_tlm = :CMD if flags & OPENC3_CMD_FLAG_MASK == OPENC3_CMD_FLAG_MASK
      stored = false
      stored = true if flags & OPENC3_STORED_FLAG_MASK == OPENC3_STORED_FLAG_MASK
      id = false
      id = true if flags & OPENC3_ID_FLAG_MASK == OPENC3_ID_FLAG_MASK
      cbor = false
      cbor = true if flags & OPENC3_CBOR_FLAG_MASK == OPENC3_CBOR_FLAG_MASK
      includes_received_time = false
      includes_received_time = true if flags & OPENC3_RECEIVED_TIME_FLAG_MASK == OPENC3_RECEIVED_TIME_FLAG_MASK
      includes_extra = false
      includes_extra = true if flags & OPENC3_EXTRA_FLAG_MASK == OPENC3_EXTRA_FLAG_MASK

      if flags & OPENC3_ENTRY_TYPE_MASK == OPENC3_JSON_PACKET_ENTRY_TYPE_MASK
        packet_index, time_nsec_since_epoch = entry[2..11].unpack('nQ>')
        received_time_nsec_since_epoch, extra, json_data = handle_received_time_extra_and_data(entry, time_nsec_since_epoch, includes_received_time, includes_extra, cbor)
        lookup_cmd_or_tlm, target_name, packet_name, _id, key_map = @packets[packet_index]
        if cmd_or_tlm != lookup_cmd_or_tlm
          raise "Packet type mismatch, packet:#{cmd_or_tlm}, lookup:#{lookup_cmd_or_tlm}"
        end

        if cbor
          return JsonPacket.new(cmd_or_tlm, target_name, packet_name, time_nsec_since_epoch, stored, CBOR.decode(json_data), key_map, received_time_nsec_since_epoch: received_time_nsec_since_epoch, extra: extra)
        else
          return JsonPacket.new(cmd_or_tlm, target_name, packet_name, time_nsec_since_epoch, stored, json_data, key_map, received_time_nsec_since_epoch: received_time_nsec_since_epoch, extra: extra)
        end
      elsif flags & OPENC3_ENTRY_TYPE_MASK == OPENC3_RAW_PACKET_ENTRY_TYPE_MASK
        packet_index, time_nsec_since_epoch = entry[2..11].unpack('nQ>')
        received_time_nsec_since_epoch, _extra, packet_data = handle_received_time_extra_and_data(entry, time_nsec_since_epoch, includes_received_time, includes_extra, cbor)
        lookup_cmd_or_tlm, target_name, packet_name, _id = @packets[packet_index]
        if cmd_or_tlm != lookup_cmd_or_tlm
          raise "Packet type mismatch, packet:#{cmd_or_tlm}, lookup:#{lookup_cmd_or_tlm}"
        end

        if identify_and_define
          packet = identify_and_define_packet_data(cmd_or_tlm, target_name, packet_name, packet_data)
        else
          # Build Packet
          packet = Packet.new(target_name, packet_name, :BIG_ENDIAN, nil, packet_data)
        end
        packet.packet_time = Time.from_nsec_from_epoch(time_nsec_since_epoch)
        received_time = Time.from_nsec_from_epoch(received_time_nsec_since_epoch)
        packet.set_received_time_fast(received_time)
        packet.cmd_or_tlm = cmd_or_tlm
        packet.stored = stored
        packet.received_count += 1
        return packet
      elsif flags & OPENC3_ENTRY_TYPE_MASK == OPENC3_TARGET_DECLARATION_ENTRY_TYPE_MASK
        target_name_length = length - OPENC3_PRIMARY_FIXED_SIZE - OPENC3_TARGET_DECLARATION_SECONDARY_FIXED_SIZE
        target_name_length -= OPENC3_ID_FIXED_SIZE if id
        target_name = entry[2..(target_name_length + 1)]
        if id
          id = entry[(target_name_length + 3)..(target_name_length + 34)]
          @target_ids << id
        end
        @target_names << target_name
        return read(identify_and_define)
      elsif flags & OPENC3_ENTRY_TYPE_MASK == OPENC3_PACKET_DECLARATION_ENTRY_TYPE_MASK
        target_index = entry[2..3].unpack('n')[0]
        target_name = @target_names[target_index]
        unless target_name
          # There was a bug in the PacketLogWriter before version 5.6.0 that stored an invalid target_index
          # Attempt to work around by reading the target_name from the filename
          filename_split = @filename.to_s.split('__')
          target_name = filename_split[3]
          target_name = 'UNKNOWN' unless target_name
        end
        packet_name_length = length - OPENC3_PRIMARY_FIXED_SIZE - OPENC3_PACKET_DECLARATION_SECONDARY_FIXED_SIZE
        packet_name_length -= OPENC3_ID_FIXED_SIZE if id
        packet_name = entry[4..(packet_name_length + 3)]
        if id
          id = entry[(packet_name_length + 4)..-1]
          @packet_ids << id
        end
        @packets << [cmd_or_tlm, target_name, packet_name, id]
        return read(identify_and_define)
      elsif flags & OPENC3_ENTRY_TYPE_MASK == OPENC3_KEY_MAP_ENTRY_TYPE_MASK
        packet_index = entry[2..3].unpack('n')[0]
        key_map_length = length - OPENC3_PRIMARY_FIXED_SIZE - OPENC3_KEY_MAP_SECONDARY_FIXED_SIZE
        if cbor
          key_map = CBOR.decode(entry[4..(key_map_length + 3)])
        else
          key_map = JSON.parse(entry[4..(key_map_length + 3)], :allow_nan => true, :create_additions => true)
        end
        @packets[packet_index] << key_map
        return read(identify_and_define)
      elsif flags & OPENC3_ENTRY_TYPE_MASK == OPENC3_OFFSET_MARKER_ENTRY_TYPE_MASK
        data = entry[2..-1]
        split_data = data.split(',')
        redis_offset = split_data[0]
        redis_topic = split_data[1]
        if redis_topic
          @last_offsets[redis_topic] = redis_offset
        else
          @redis_offset = redis_offset
        end
        return read(identify_and_define)
      else
        raise "Invalid Entry Flags: #{flags}"
      end
    rescue => e
      close()
      raise e
    end

    # @return [Integer] The size of the log file being processed
    def size
      @file.stat.size
    end

    # @return [Integer] The current file position in the log file
    def bytes_read
      @file.pos
    end

    protected

    def reset
      @file = nil
      @filename = nil
      @max_read_size = MAX_READ_SIZE
      @target_names = []
      @target_ids = []
      @packets = []
      @packet_ids = []
      @redis_offset = nil
      @last_offsets = {}
    end

    # This is best effort. May return unidentified/undefined packets
    def identify_and_define_packet_data(cmd_or_tlm, target_name, packet_name, packet_data)
      packet = nil
      unless target_name and packet_name
        if cmd_or_tlm == :CMD
          packet = System.commands.identify(packet_data)
        else
          packet = System.telemetry.identify!(packet_data)
        end
      else
        begin
          if cmd_or_tlm == :CMD
            packet = System.commands.packet(target_name, packet_name)
          else
            packet = System.telemetry.packet(target_name, packet_name)
          end
          packet.buffer = packet_data
        rescue
          # Could not find a definition for this packet
          Logger.instance.error "Unknown packet #{target_name} #{packet_name}"
          packet = Packet.new(target_name, packet_name, :BIG_ENDIAN, nil, packet_data)
        end
      end
      packet
    end

    # Should return if successfully switched to requested configuration
    def read_file_header
      header = @file.read(OPENC3_HEADER_LENGTH)
      if header and header.length == OPENC3_HEADER_LENGTH
        if header == OPENC3_FILE_HEADER
          # Found OpenC3 5 File Header - That's all we need to do
        elsif header == COSMOS4_FILE_HEADER
          raise "COSMOS 4 log file must be converted to OpenC3 5"
        elsif header == COSMOS2_FILE_HEADER
          raise "COSMOS 2 log file must be converted to OpenC3 5"
        else
          raise "OpenC3 file header not found"
        end
      else
        raise "Failed to read at least #{OPENC3_HEADER_LENGTH} bytes from packet log"
      end
    end

    # Handle common optional fields in raw and JSON packets
    def handle_received_time_extra_and_data(entry, time_nsec_since_epoch, includes_received_time, includes_extra, cbor)
      next_offset = 12
      received_time_nsec_since_epoch = time_nsec_since_epoch
      if includes_received_time
        received_time_nsec_since_epoch = entry[next_offset..(next_offset + OPENC3_RECEIVED_TIME_FIXED_SIZE - 1)].unpack(OPENC3_RECEIVED_TIME_PACK_DIRECTIVE)[0]
        next_offset += OPENC3_RECEIVED_TIME_FIXED_SIZE
      end
      extra = nil
      if includes_extra
        extra_length = entry[next_offset..(next_offset + OPENC3_EXTRA_LENGTH_FIXED_SIZE - 1)].unpack(OPENC3_EXTRA_LENGTH_PACK_DIRECTIVE)[0]
        next_offset += OPENC3_EXTRA_LENGTH_FIXED_SIZE
        extra_encoded = entry[next_offset..(next_offset + extra_length - 1)]
        next_offset += extra_length
        if cbor
          extra = CBOR.decode(extra_encoded)
        else
          extra = JSON.parse(extra_encode, allow_nan: true, create_additions: true)
        end
      end
      data = entry[next_offset..-1]
      return received_time_nsec_since_epoch, extra, data
    end
  end
end

File: openc3/python/openc3/logs/packet_log_reader.py
# Copyright 2025 OpenC3, Inc.
# All Rights Reserved.
#
# This program is free software; you can modify and/or redistribute it
# under the terms of the GNU Affero General Public License
# as published by the Free Software Foundation; version 3 with
# attribution addendums as found in the LICENSE.txt
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU Affero General Public License for more details.
#
# This file may also be used under the terms of a commercial license
# if purchased from OpenC3, Inc.

"""
Packet log reader for COSMOS5 binary log files.

Reads packet logs containing either commands or telemetry in raw or JSON format.
This is a port of the Ruby PacketLogReader class.
"""

import json
import struct
from collections.abc import Iterator
from datetime import datetime, timezone

import cbor2

from openc3.logs.packet_log_constants import (
    COSMOS2_FILE_HEADER,
    COSMOS4_FILE_HEADER,
    OPENC3_CBOR_FLAG_MASK,
    OPENC3_CMD_FLAG_MASK,
    OPENC3_ENTRY_TYPE_MASK,
    OPENC3_EXTRA_FLAG_MASK,
    OPENC3_EXTRA_LENGTH_FIXED_SIZE,
    OPENC3_FILE_HEADER,
    OPENC3_HEADER_LENGTH,
    OPENC3_ID_FIXED_SIZE,
    OPENC3_ID_FLAG_MASK,
    OPENC3_JSON_PACKET_ENTRY_TYPE_MASK,
    OPENC3_KEY_MAP_ENTRY_TYPE_MASK,
    OPENC3_KEY_MAP_SECONDARY_FIXED_SIZE,
    OPENC3_OFFSET_MARKER_ENTRY_TYPE_MASK,
    OPENC3_PACKET_DECLARATION_ENTRY_TYPE_MASK,
    OPENC3_PACKET_DECLARATION_SECONDARY_FIXED_SIZE,
    OPENC3_PRIMARY_FIXED_SIZE,
    OPENC3_RAW_PACKET_ENTRY_TYPE_MASK,
    OPENC3_RECEIVED_TIME_FIXED_SIZE,
    OPENC3_RECEIVED_TIME_FLAG_MASK,
    OPENC3_STORED_FLAG_MASK,
    OPENC3_TARGET_DECLARATION_ENTRY_TYPE_MASK,
    OPENC3_TARGET_DECLARATION_SECONDARY_FIXED_SIZE,
)
from openc3.packets.json_packet import JsonPacket
from openc3.packets.packet import Packet


class PacketLogReader:
    """
    Reads a packet log of either commands or telemetry.

    Supports both raw packet data and JSON (decommutated) packet data.
    """

    MAX_READ_SIZE = 1_000_000_000

    def __init__(self):
        """Create a new log file reader."""
        self._reset()

    def _reset(self):
        """Reset internal state for a new file."""
        self._file = None
        self._filename = None
        self._max_read_size = self.MAX_READ_SIZE
        self._target_names = []
        self._target_ids = []
        self._packets = []  # List of [cmd_or_tlm, target_name, packet_name, id, key_map]
        self._packet_ids = []
        self._redis_offset = None
        self._last_offsets = {}

    @property
    def redis_offset(self) -> str | None:
        """The Redis offset from the log file."""
        return self._redis_offset

    @property
    def last_offsets(self) -> dict:
        """Dictionary of topic -> offset from the log file."""
        return self._last_offsets

    @property
    def filename(self) -> str | None:
        """The currently open filename."""
        return self._filename

    def each(
        self,
        filename: str,
        identify_and_define: bool = True,
        start_time: datetime | None = None,
        end_time: datetime | None = None,
    ) -> Iterator[Packet | JsonPacket]:
        """
        Iterate over each packet in the log file.

        Args:
            filename: The log file to read
            identify_and_define: Whether to identify and define raw packets
            start_time: Optional start time filter (packets before are skipped)
            end_time: Optional end time filter (packets after cause iteration to stop)

        Yields:
            Packet or JsonPacket objects

        Returns:
            True if end_time was reached, False otherwise (via StopIteration value)
        """
        reached_end_time = False
        try:
            self.open(filename)

            while True:
                packet = self.read(identify_and_define)
                if packet is None:
                    break

                time = packet.packet_time
                if time:
                    if start_time and time < start_time:
                        continue
                    if end_time and time > end_time:
                        reached_end_time = True
                        break

                yield packet

        finally:
            self.close()

        return reached_end_time

    def open(self, filename: str):
        """
        Open a log file for reading.

        Args:
            filename: Path to the log file

        Raises:
            ValueError: If the file header is invalid
            IOError: If the file cannot be read
        """
        self.close()
        self._reset()
        self._filename = filename
        self._file = open(filename, "rb")  # noqa: SIM115

        file_size = self._file.seek(0, 2)  # Seek to end to get size
        self._file.seek(0)  # Seek back to start

        self._max_read_size = min(file_size, self.MAX_READ_SIZE)
        self._read_file_header()

    def close(self):
        """Close the current log file."""
        if self._file and not self._file.closed:
            self._file.close()
        self._file = None

    def read(self, identify_and_define: bool = True) -> Packet | JsonPacket | None:
        """
        Read a packet from the log file.

        Args:
            identify_and_define: Whether to identify and define raw packets
                using the System definitions

        Returns:
            Packet or JsonPacket, or None if end of file
        """
        # Read entry length (4 bytes, big-endian)
        length_bytes = self._file.read(4)
        if not length_bytes or len(length_bytes) < 4:
            return None

        length = struct.unpack(">I", length_bytes)[0]
        entry = self._file.read(length)
        if len(entry) < length:
            return None

        # Parse flags (2 bytes, big-endian)
        flags = struct.unpack(">H", entry[0:2])[0]

        cmd_or_tlm = "CMD" if flags & OPENC3_CMD_FLAG_MASK else "TLM"
        stored = bool(flags & OPENC3_STORED_FLAG_MASK)
        has_id = bool(flags & OPENC3_ID_FLAG_MASK)
        is_cbor = bool(flags & OPENC3_CBOR_FLAG_MASK)
        has_extra = bool(flags & OPENC3_EXTRA_FLAG_MASK)
        has_received_time = bool(flags & OPENC3_RECEIVED_TIME_FLAG_MASK)

        entry_type = flags & OPENC3_ENTRY_TYPE_MASK

        if entry_type == OPENC3_JSON_PACKET_ENTRY_TYPE_MASK:
            return self._read_json_packet(entry, cmd_or_tlm, stored, is_cbor, has_received_time, has_extra)

        elif entry_type == OPENC3_RAW_PACKET_ENTRY_TYPE_MASK:
            return self._read_raw_packet(
                entry, cmd_or_tlm, stored, is_cbor, has_received_time, has_extra, identify_and_define
            )

        elif entry_type == OPENC3_TARGET_DECLARATION_ENTRY_TYPE_MASK:
            self._read_target_declaration(entry, length, has_id)
            return self.read(identify_and_define)

        elif entry_type == OPENC3_PACKET_DECLARATION_ENTRY_TYPE_MASK:
            self._read_packet_declaration(entry, length, cmd_or_tlm, has_id)
            return self.read(identify_and_define)

        elif entry_type == OPENC3_KEY_MAP_ENTRY_TYPE_MASK:
            self._read_key_map(entry, length, is_cbor)
            return self.read(identify_and_define)

        elif entry_type == OPENC3_OFFSET_MARKER_ENTRY_TYPE_MASK:
            self._read_offset_marker(entry)
            return self.read(identify_and_define)

        else:
            raise ValueError(f"Invalid Entry Flags: {flags:#06x}")

    @property
    def size(self) -> int:
        """The size of the log file being processed."""
        if self._file:
            pos = self._file.tell()
            self._file.seek(0, 2)
            size = self._file.tell()
            self._file.seek(pos)
            return size
        return 0

    @property
    def bytes_read(self) -> int:
        """The current file position in the log file."""
        if self._file:
            return self._file.tell()
        return 0

    def _read_file_header(self):
        """Read and validate the file header."""
        header = self._file.read(OPENC3_HEADER_LENGTH)

        if not header or len(header) < OPENC3_HEADER_LENGTH:
            raise ValueError(f"Failed to read at least {OPENC3_HEADER_LENGTH} bytes from packet log")

        if header == OPENC3_FILE_HEADER:
            # Found OpenC3 5 File Header - valid
            pass
        elif header == COSMOS4_FILE_HEADER:
            raise ValueError("COSMOS 4 log file must be converted to OpenC3 5")
        elif header == COSMOS2_FILE_HEADER:
            raise ValueError("COSMOS 2 log file must be converted to OpenC3 5")
        else:
            raise ValueError("OpenC3 file header not found")

    def _handle_received_time_extra_and_data(
        self, entry: bytes, time_nsec: int, has_received_time: bool, has_extra: bool, is_cbor: bool
    ):
        """
        Handle optional received_time and extra fields, then return the data portion.

        Args:
            entry: The full entry bytes
            time_nsec: The packet timestamp in nanoseconds
            has_received_time: Whether entry includes received_time
            has_extra: Whether entry includes extra data
            is_cbor: Whether data is CBOR encoded

        Returns:
            Tuple of (received_time_nsec, extra, data_bytes)
        """
        next_offset = 12  # After flags (2) + packet_index (2) + timestamp (8)
        received_time_nsec = time_nsec

        if has_received_time:
            received_time_nsec = struct.unpack(">Q", entry[next_offset : next_offset + 8])[0]
            next_offset += OPENC3_RECEIVED_TIME_FIXED_SIZE

        extra = None
        if has_extra:
            extra_length = struct.unpack(">I", entry[next_offset : next_offset + 4])[0]
            next_offset += OPENC3_EXTRA_LENGTH_FIXED_SIZE
            extra_encoded = entry[next_offset : next_offset + extra_length]
            next_offset += extra_length

            if is_cbor:
                extra = cbor2.loads(extra_encoded)
            else:
                extra = json.loads(extra_encoded)

        data = entry[next_offset:]
        return received_time_nsec, extra, data

    def _read_json_packet(
        self, entry: bytes, cmd_or_tlm: str, stored: bool, is_cbor: bool, has_received_time: bool, has_extra: bool
    ) -> JsonPacket:
        """Read a JSON packet entry."""
        # Parse packet_index (2 bytes) and timestamp (8 bytes)
        packet_index, time_nsec = struct.unpack(">HQ", entry[2:12])

        received_time_nsec, extra, json_data = self._handle_received_time_extra_and_data(
            entry, time_nsec, has_received_time, has_extra, is_cbor
        )

        if packet_index >= len(self._packets):
            raise ValueError(f"Invalid packet_index {packet_index}, only {len(self._packets)} packets declared")

        packet_info = self._packets[packet_index]
        lookup_cmd_or_tlm, target_name, packet_name, _id = packet_info[:4]
        key_map = packet_info[4] if len(packet_info) > 4 else None

        if cmd_or_tlm != lookup_cmd_or_tlm:
            raise ValueError(f"Packet type mismatch, packet:{cmd_or_tlm}, lookup:{lookup_cmd_or_tlm}")

        if is_cbor:
            json_hash = cbor2.loads(json_data)
        else:
            # Decode bytes to string for JsonPacket to parse
            json_hash = json_data.decode("utf-8") if isinstance(json_data, bytes) else json_data

        return JsonPacket(
            cmd_or_tlm,
            target_name,
            packet_name,
            time_nsec,
            stored,
            json_hash,
            key_map,
            received_time_nsec_since_epoch=received_time_nsec,
            extra=extra,
        )

    def _read_raw_packet(
        self,
        entry: bytes,
        cmd_or_tlm: str,
        stored: bool,
        is_cbor: bool,
        has_received_time: bool,
        has_extra: bool,
        identify_and_define: bool,
    ) -> Packet:
        """Read a raw packet entry."""
        # Parse packet_index (2 bytes) and timestamp (8 bytes)
        packet_index, time_nsec = struct.unpack(">HQ", entry[2:12])

        received_time_nsec, extra, packet_data = self._handle_received_time_extra_and_data(
            entry, time_nsec, has_received_time, has_extra, is_cbor
        )

        if packet_index >= len(self._packets):
            raise ValueError(f"Invalid packet_index {packet_index}, only {len(self._packets)} packets declared")

        packet_info = self._packets[packet_index]
        lookup_cmd_or_tlm, target_name, packet_name, _id = packet_info[:4]

        if cmd_or_tlm != lookup_cmd_or_tlm:
            raise ValueError(f"Packet type mismatch, packet:{cmd_or_tlm}, lookup:{lookup_cmd_or_tlm}")

        if identify_and_define:
            packet = self._identify_and_define_packet_data(cmd_or_tlm, target_name, packet_name, packet_data)
        else:
            packet = Packet(target_name, packet_name, "BIG_ENDIAN", None, packet_data)

        packet.packet_time = self._nsec_to_datetime(time_nsec)
        packet.received_time = self._nsec_to_datetime(received_time_nsec)
        packet.cmd_or_tlm = cmd_or_tlm
        packet.stored = stored
        packet.extra = extra
        # Store original nanosecond timestamps for high-precision access
        packet._time_nsec = time_nsec
        packet._received_time_nsec = received_time_nsec
        packet.received_count += 1

        return packet

    def _read_target_declaration(self, entry: bytes, length: int, has_id: bool):
        """Read a target declaration entry."""
        target_name_length = length - OPENC3_PRIMARY_FIXED_SIZE - OPENC3_TARGET_DECLARATION_SECONDARY_FIXED_SIZE
        if has_id:
            target_name_length -= OPENC3_ID_FIXED_SIZE

        target_name = entry[2 : 2 + target_name_length].decode("utf-8")

        if has_id:
            target_id = entry[2 + target_name_length : 2 + target_name_length + OPENC3_ID_FIXED_SIZE]
            self._target_ids.append(target_id)

        self._target_names.append(target_name)

    def _read_packet_declaration(self, entry: bytes, length: int, cmd_or_tlm: str, has_id: bool):
        """Read a packet declaration entry."""
        target_index = struct.unpack(">H", entry[2:4])[0]

        if target_index < len(self._target_names):
            target_name = self._target_names[target_index]
        else:
            # Workaround for bug in PacketLogWriter before version 5.6.0
            # Try to extract target name from filename
            if self._filename:
                filename_parts = self._filename.split("__")
                target_name = filename_parts[3] if len(filename_parts) > 3 else "UNKNOWN"
            else:
                target_name = "UNKNOWN"

        packet_name_length = length - OPENC3_PRIMARY_FIXED_SIZE - OPENC3_PACKET_DECLARATION_SECONDARY_FIXED_SIZE
        if has_id:
            packet_name_length -= OPENC3_ID_FIXED_SIZE

        packet_name = entry[4 : 4 + packet_name_length].decode("utf-8")

        packet_id = None
        if has_id:
            packet_id = entry[4 + packet_name_length :]
            self._packet_ids.append(packet_id)

        # Store packet info (key_map will be appended later if KEY_MAP entry follows)
        self._packets.append([cmd_or_tlm, target_name, packet_name, packet_id])

    def _read_key_map(self, entry: bytes, length: int, is_cbor: bool):
        """Read a key map entry for CBOR compressed packets."""
        packet_index = struct.unpack(">H", entry[2:4])[0]
        key_map_length = length - OPENC3_PRIMARY_FIXED_SIZE - OPENC3_KEY_MAP_SECONDARY_FIXED_SIZE

        key_map_bytes = entry[4 : 4 + key_map_length]

        if is_cbor:
            key_map = cbor2.loads(key_map_bytes)
        else:
            key_map = json.loads(key_map_bytes)

        if packet_index < len(self._packets):
            # Append key_map to existing packet entry
            if len(self._packets[packet_index]) == 4:
                self._packets[packet_index].append(key_map)
            else:
                self._packets[packet_index][4] = key_map

    def _read_offset_marker(self, entry: bytes):
        """Read an offset marker entry."""
        data = entry[2:].decode("utf-8")
        parts = data.split(",")
        redis_offset = parts[0]

        if len(parts) > 1:
            redis_topic = parts[1]
            self._last_offsets[redis_topic] = redis_offset
        else:
            self._redis_offset = redis_offset

    def _identify_and_define_packet_data(
        self, cmd_or_tlm: str, target_name: str, packet_name: str, packet_data: bytes
    ) -> Packet:
        """
        Identify and define a packet using System definitions.

        This is best effort - may return unidentified/undefined packets.
        """
        packet = None

        if target_name and packet_name:
            try:
                from openc3.system.system import System

                if cmd_or_tlm == "CMD":
                    packet = System.commands.packet(target_name, packet_name)
                else:
                    packet = System.telemetry.packet(target_name, packet_name)
                packet.buffer = packet_data
            except Exception:
                # Could not find a definition for this packet
                from openc3.utilities.logger import Logger

                Logger.error(f"Unknown packet {target_name} {packet_name}")
                packet = Packet(target_name, packet_name, "BIG_ENDIAN", None, packet_data)
        else:
            try:
                from openc3.system.system import System

                if cmd_or_tlm == "CMD":
                    packet = System.commands.identify(packet_data)
                else:
                    packet = System.telemetry.identify(packet_data)
            except Exception:
                packet = Packet(target_name, packet_name, "BIG_ENDIAN", None, packet_data)

        if packet is None:
            packet = Packet(target_name, packet_name, "BIG_ENDIAN", None, packet_data)

        return packet

    @staticmethod
    def _nsec_to_datetime(nsec: int) -> datetime:
        """Convert nanoseconds since epoch to datetime."""
        seconds = nsec / 1_000_000_000
        return datetime.fromtimestamp(seconds, tz=timezone.utc)
